### PR TITLE
remove exception raising from noexcept handlers

### LINKED ISSUE #82: handler functions should not raise python exceptions
Copied from a comment of @Nikratio in #81 : 

> The `cdef void` functions in `handlers.pxi` that are assigned to `fuse_lowlevel_ops` should never raise exceptions. Where do you see them use `raise`?

Nevermind, found it. I believe the affected methods are `fuse_write` and `fuse_setxattr`  (which raise `OverflowError` if the data size doesn't fit into `ssize_t` (pretty unlikely) and `fuse_init` (which raises a `RuntimeError` if the kernel is too old). 

Neither of these checks can easily be moved elsewhere. For the first two, I would suggest to instead return EINVAL through `fuse_reply_err()` and log the error (using Python's logging module). For the third, the best I can think of is to set some sort of global flag, check it in `init()` (after the call to fuse_session_new) and raise the exception there if set.

### PR DESCRIPTION
**Pull request by ThomasWaldmann**
fixes #82

### CODE REVIEW

**Comment by ThomasWaldmann** [comment]
@Nikratio like that?

What do we do with that one? Also log + fuse_reply_err with EINVAL?

```
    # Make sure we know all the flags
    if c.flags & ~(libc_extra.XATTR_CREATE | libc_extra.XATTR_REPLACE):
        raise ValueError('unknown flag(s): %o' % c.flags)
```

File: src/handlers.pxi [hunk 1]
```diff
@@ -689,7 +699,11 @@ async def fuse_setxattr_async (_Container c, name, value):
 
     # Make sure we know all the flags
     if c.flags & ~(libc_extra.XATTR_CREATE | libc_extra.XATTR_REPLACE):
-        raise ValueError('unknown flag(s): %o' % c.flags)
```
Comment (Nikratio) [inline src/handlers.pxi:692]: This is fine. `fuse_setxattr_async` is not called directly through libfuse, so raising an exception here will be handled correctly.

**Comment by Nikratio** [comment]
Note that the `async` functions are not evaluated when called by the C request handers. For example, ` save_retval(fuse_setxattr_async(c, name, value))` in `fuse_setxattr` does not evaluate the code in `fuse_setxattr_async` but instead generates a closure that is stored in the global `py_retval` object and evaluated by the (Python) code in https://github.com/libfuse/pyfuse3/blob/master/src/internal.pxi#L272. At that point, any raised exceptions will be handled correctly.

**Comment by Nikratio** [comment]
Thinking about this some more, maybe an alternative solution would be to declare https://github.com/libfuse/pyfuse3/blob/master/Include/fuse_lowlevel.pxd#L184 as "except?" so that Cython checks for pending exceptions whenever this function returns. At that point, it would also pick-up any exceptions that were raised by the C-level callbacks.

This might be worth testing, if we can find a way to make Cython compile the code without complaints.

File: src/handlers.pxi (hunk 1 above)
Comment (ThomasWaldmann) [reply src/handlers.pxi]: thanks, i removed that changeset again.

**Comment by ThomasWaldmann** [comment]
`fuse_session_process_buf` has void return type, so guess we can't use `except? <somevalue>`.  `except *` might be an option.

I don't really understand how this all works together, so maybe that should be done by someone who does.

**Comment by Nikratio** [comment]
Python exceptions are stored in separate, per-thread variables. The return value of a C function just indicates whether everything went right, or if an exception was stored.

For functions that do not return a value, Cython has the `except?` specifier. This means that Cython will check the exception values after *every* call (thus incurring a tiny performance overhead).

I can look into this, you're just way faster with pull requests than me :-).

**Comment by ThomasWaldmann** [comment]
Guess you mean `except *`.
The cython docs only mention `except? <returnvalue>`.

**Comment by Nikratio** [comment]
Looked into this - alternate patch in issue #82

**Comment by ThomasWaldmann** [comment]
Closed in favour of #86.

**Pull request closed without merging**